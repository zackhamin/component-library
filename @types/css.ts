export type Unit = 'px' | 'rem' | 'em' | '%' | 'vh' | 'vw';
export type AngleUnit = 'deg' | 'rad' | 'grad' | 'turn';
export type TextAlign = 'left' | 'center' | 'right';

export type Size = `${number}${Unit}` | '0';

export type Rotation = `${number}${AngleUnit}`;
