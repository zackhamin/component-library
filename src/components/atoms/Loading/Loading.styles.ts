import { colors } from 'Styles';
import styled from 'styled-components';

const StyledLoader = styled.div<{
	color: string;
	size: string;
}>`
		font-size: 10px;
		margin: 5rem auto;
		text-indent: -9999em;
		width: ${({ size }) => size};
		height: ${({ size }) => size};
		border-radius: 50%;
		background: ${({ color }) => color};
		background: linear-gradient(
			to right,
			${({ color }) => color} 10%,
			rgba(255, 255, 255, 0) 42%
		);
		position: relative;
		animation: load 1.4s infinite linear;
		transform: translateZ(0);
	}

	:before {
		width: 50%;
		height: 50%;
		background: ${({ color }) => color};
		border-radius: 100% 0 0 0;
		position: absolute;
		top: 0;
		left: 0;
		content: '';
	}

	:after {
		background: ${colors.white};
		width: 75%;
		height: 75%;
		border-radius: 50%;
		content: '';
		margin: auto;
		position: absolute;
		top: 0;
		left: 0;
		bottom: 0;
		right: 0;
	}

	@keyframes load {
		0% {
			transform: rotate(0deg);
		}
		100% {
			transform: rotate(360deg);
		}
`;

export { StyledLoader };
