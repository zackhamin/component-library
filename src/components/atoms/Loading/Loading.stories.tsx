import { colors } from 'src/styles';
import { Size } from '@Types/css';
import { Loading, LoadingProps } from './Loading';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'atoms/Loading',
	component: Loading,
	argTypes: {
		children: {
			description: 'text read out by screen reader',
			control: {
				type: 'text',
			},
		},
		color: {
			control: {
				type: 'color',
			},
		},
		size: {
			control: {
				type: 'text',
			},
		},
	},
} as Meta;

const Template: Story<LoadingProps> = (args) => <Loading {...args} />;

const defaultProps = {
	children: 'Loading',
	color: colors.emerald,
	size: '6.4rem' as Size,
};

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};
