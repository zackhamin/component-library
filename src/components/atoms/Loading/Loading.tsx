import { colors } from 'Styles';
import { Size } from '@Types/css';
import { StyledLoader } from './Loading.styles';

export interface LoadingProps {
	children?: React.ReactNode;
	color?: string;
	size?: Size;
}

const Loading = ({
	children = 'Loading',
	color = colors.emerald,
	size = '6.4rem',
}: LoadingProps) => {
	return (
		<StyledLoader color={color} size={size}>
			{children}
		</StyledLoader>
	);
};

export { Loading };
