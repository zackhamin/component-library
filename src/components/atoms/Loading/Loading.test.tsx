import { Loading } from '../Loading';
import { Size } from '@Types/css';
import { render, screen } from '@testing-library/react';

describe('Loading', () => {
	it('renders correctly', () => {
		const { container } = render(<Loading />);
		expect(container).toMatchSnapshot();
	});

	it('renders correctly with props', () => {
		const props = {
			children: 'children',
			color: 'black',
			size: '1em' as Size,
		};
		const { container } = render(<Loading {...props} />);

		const loading = screen.getByText(props.children);

		expect(loading).toBeInTheDocument();
		expect(loading).toHaveStyle({
			width: props.size,
			height: props.size,
		});
		expect(loading.getAttribute('color')).toBe(props.color);
	});
});
