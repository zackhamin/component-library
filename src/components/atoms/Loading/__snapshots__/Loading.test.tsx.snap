// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading renders correctly 1`] = `
.c0 {
  font-size: 10px;
  margin: 5rem auto;
  text-indent: -9999em;
  width: 6.4rem;
  height: 6.4rem;
  border-radius: 50%;
  background: #00594C;
  background: linear-gradient( to right, #00594C 10%, rgba(255,255,255,0) 42% );
  position: relative;
  -webkit-animation: load 1.4s infinite linear;
  animation: load 1.4s infinite linear;
  -webkit-transform: translateZ(0);
  -ms-transform: translateZ(0);
  transform: translateZ(0);
}

.c0:before {
  width: 50%;
  height: 50%;
  background: #00594C;
  border-radius: 100% 0 0 0;
  position: absolute;
  top: 0;
  left: 0;
  content: '';
}

.c0:after {
  background: #ffffff;
  width: 75%;
  height: 75%;
  border-radius: 50%;
  content: '';
  margin: auto;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

<div>
  <div
    class="c0"
    color="#00594C"
  >
    Loading
  </div>
</div>
`;
